designdocs/provider_registration.md

```diff
@@ -0,0 +1,175 @@
+1. The provider's MSA id is added to storage on chain.
+
+Should a provider choose at any point to stop providing services, it can issue a
+request to remove itself from the provider registry:
+1. The provider issues an announcement that it will stop providing services.
+1. The provider's MSA id is removed from storage.
+
+
+Please note:
+* All names are placeholders and may be changed.
+* Types may change as needed during implementation phase
+* Errors in the extrinsic(s) must have different, reasonably-named error enums
+  for each type of error for ease of debugging.
+
+### Types
+* `ProviderRegistrationParams<T: Config>`, the arguments used to register an announcement.
+  * `provider_msa_id`: `MsaId`
+  * `provider_metadata`: `ProviderAnnouncementMetadata`
+* `ProviderRegistrationAnnouncement<T: Config>`, the resource that exists on-chain
+  * `block_number`: `BlockNumber`
+  * `provider_msa_id`: `MsaId`
+  * `provider_metadata`: `ProviderAnnouunucementMetadata`
+* `ProviderUnregistrationParams<T:Config>`, the arguments used to unregister an announcement.
+  * `provider_msa_id`: `MsaId`
+* `ProviderUnregistrationAnnouncement<T:Config>`, the resource that exists on-chain
+  * `provider_msa_id`: `MsaId`
+* `ProviderAnnouncementMetadata`
+  * `name`: `Vec<u8>`
+* `Provider`
+  * `provider_msa_id`: `MsaId`
+  * `name`: `Vec<u8>`
+* `ProviderRegistrationAnnouncementOptions<T:Config>`
+  * `provider_msa_id`:  `Option<MsaId>`, the announcer's MSA id.  Pass None() to get all announcements.
+  * `from_block`: `<T::BlockNumber>`, retrieve messages starting at the given block number (inclusive)
+  * `to_block`: `<T::BlockNumber>`, retrieve messages ending at the given block number (inclusive)
+  * `from_index`: `u32`, starting message index
+  * `page_size`: `usize`, retrieve `page_size` messages at a time, up to configured `T::PageSizeMax`. If 0, return `T::PageSizeMax` results
+
+#### Storage
+* `ProviderRegistry<T: Config>`: `StorageMap<MsaId, Provider>`
+  * Stores registered providers and provides lookup functionality via `MsaId`.
+    Existence in this storage structure implies that a provider's fee has been
+    paid and their registration was otherwise successful.
+
+### Extrinsics
+#### register_provider(origin, registration_params)
+Creates and posts a `ProviderRegistrationAnnouncement` on chain. The `MsaId`
+included in the announcement must already exist.
+
+This extrinsic is responsible for storing the registered provider in the
+`ProviderRegistry`.
+
+* **Parameters**
+  * `origin`: `Origin`  required for all extrinsics, the caller/sender.
+  * `registration_params`: `ProviderRegistrationParams`, the parameters to use in the registration announcement.
+* **Event**:  `Event::<T>::ProviderRegistered(provider_msa_id, provider_metadata)`
+* **Restrictions**:
+  * `origin`'s `msa_id` must have capacity to post the transaction (including fee) during the current epoch
+
+#### unregister_provider(origin, unregistration_params)
```

Comment:
nit: for some reason using unregister is a bit off to me. what about `remove_provider`?

Reply:
i see what you mean. Another possibilty: what about `deregister_provider`?